dp-hello-world-event/service/service.go

```diff
@@ -0,0 +1,165 @@
+	shutdownTimeout time.Duration
+}
+
+// Run the service
+func Run(ctx context.Context, serviceList *ExternalServiceList, buildTime, gitCommit, version string, svcErrors chan error) (*Service, error) {
+	log.Event(ctx, "running service", log.INFO)
+
+	// Read config
+	cfg, err := config.Get()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to retrieve service configuration")
+	}
+	log.Event(ctx, "got service configuration", log.Data{"config": cfg}, log.INFO)
+
+	// Get HTTP Server with collectionID checkHeader middleware
+	r := mux.NewRouter()
+	s := serviceList.GetHTTPServer(cfg.BindAddr, r)
+
+	// Get Kafka consumer
+	consumer, err := serviceList.GetKafkaConsumer(ctx, cfg)
+	if err != nil {
+		log.Event(ctx, "failed to initialise kafka consumer", log.FATAL, log.Error(err))
+		return nil, err
+	}
+
+	// Event Handler for Kafka Consumer with the created S3 Clients and Vault
+	eventConsumer := event.NewConsumer()
+	eventConsumer.Consume(ctx, consumer, &event.HelloCalledHandler{})
+
+	// Get HealthCheck
+	hc, err := serviceList.GetHealthCheck(cfg, buildTime, gitCommit, version)
+	if err != nil {
+		log.Event(ctx, "could not instantiate healthcheck", log.FATAL, log.Error(err))
+		return nil, err
+	}
+
+	if err := registerCheckers(ctx, hc, consumer); err != nil {
+		return nil, errors.Wrap(err, "unable to register checkers")
+	}
+
+	r.StrictSlash(true).Path("/health").HandlerFunc(hc.Handler)
+	hc.Start(ctx)
+
+	// Run the http server in a new go-routine
+	go func() {
+		if err := s.ListenAndServe(); err != nil {
+			svcErrors <- errors.Wrap(err, "failure in http listen and serve")
+		}
+	}()
+
+	return &Service{
+		server:          s,
+		router:          r,
+		serviceList:     serviceList,
+		healthCheck:     hc,
+		consumer:        consumer,
+		eventConsumer:   eventConsumer,
+		shutdownTimeout: cfg.GracefulShutdownTimeout,
+	}, nil
+}
+
+// Close gracefully shuts the service down in the required order, with timeout
+func (svc *Service) Close(ctx context.Context) error {
+	timeout := svc.shutdownTimeout
+	log.Event(ctx, "commencing graceful shutdown", log.Data{"graceful_shutdown_timeout": timeout}, log.INFO)
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+
+	// track shutdown gracefully closes up
+	var gracefulShutdown bool
+
+	go func() {
+		defer cancel()
+		var hasShutdownError bool
+
+		// stop healthcheck, as it depends on everything else
+		if svc.serviceList.HealthCheck {
+			svc.healthCheck.Stop()
+		}
+
+		// If kafka consumer exists, stop listening to it. (Will close later)
+		if svc.serviceList.KafkaConsumer {
+			log.Event(ctx, "stopping kafka consumer listener", log.INFO)
+			if err := svc.consumer.StopListeningToConsumer(ctx); err != nil {
+				log.Event(ctx, "error stopping kafka consumer listener", log.ERROR, log.Error(err))
+				hasShutdownError = true
+			}
+			log.Event(ctx, "stopped kafka consumer listener", log.INFO)
+		}
+
+		// Close EventConsumer
+		if err := svc.eventConsumer.Close(ctx); err != nil {
```

Comment:
If we implement the described changes in the eventConsumer, we wouldn't need to close it (the consume loop go-routine will exit once the kafka Upstream channel is closed, which happens during kafkaConsumer.Close())